llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -456,15 +537,23 @@ int main(int argc, char **argv) {
     Error Err = Table.addColumn(COL_CODE, Files);
     CHECK_AND_EXIT(Err);
   }
-  if (DoSpecConst && SetSpecConstAtRT) {
-    // extract spec constant maps per each module
-    for (auto &MUptr : ResultModules) {
-      ResultSpecIDMaps.emplace_back(SpecIDMapTy());
-      if (SpecConstsMet)
-        SpecConstantsPass::collectSpecConstantMetadata(*MUptr.get(),
-                                                       ResultSpecIDMaps.back());
+  {
+    // Device library req mask is collected and stored in device image property
+    // as default and each device image module will have one req mask.
+    std::vector<uint32_t> DeviceLibReqMaskVec;
+    getDeviceLibReqMasks(ResultModules, DeviceLibReqMaskVec);
+    if (DoSpecConst && SetSpecConstAtRT) {
+      // extract spec constant maps per each module
+      for (auto &MUptr : ResultModules) {
+        ResultSpecIDMaps.emplace_back(SpecIDMapTy());
+        if (SpecConstsMet)
+          SpecConstantsPass::collectSpecConstantMetadata(
+              *MUptr.get(), ResultSpecIDMaps.back());
+      }
+      assert(DeviceLibReqMaskVec.size() == ResultSpecIDMaps.size());
     }
-    string_vector Files = saveSpecConstantIDMaps(ResultSpecIDMaps);
+    string_vector Files =
+        saveDeviceImageProperty(DeviceLibReqMaskVec, ResultSpecIDMaps);
```

Comment:
I suggest to refactor property saving now, since with addition of the lib mask property (1) it is no longer scalable and the logic becomes too complex, IMO and (2) .prop file is now generated always.

e.g. 
```
  for (unsigned I = 0; I < ResultModules.size(); ++I) {
    std::string SCFile = makeResultFileName(".prop", I);
    llvm::util::PropertySetRegistry PropSet;
    std::error_code EC;
    raw_fd_ostream SCOut(SCFile, EC);

    if (DoSpecConst && SetSpecConstAtRT) {
      PropSet.add(llvm::util::PropertySetRegistry::SYCL_SPECIALIZATION_CONSTANTS,...);
      PropSet.write(SCOut);
    }
    if (needToAddLibMask) {
      PropSet.add(llvm::util::PropertySetRegistry::SYCL_LIB_MASK, ...);
      PropSet.write(SCOut);
    }
    if (needOtherFancyFeature) {
      PropSet.add(llvm::util::PropertySetRegistry::SYCL_OTHER_FANCY_FEATURE, ...);
      PropSet.write(SCOut);
    }
    Res.emplace_back(std::move(SCFile));
  }
```

Reply:
Hi, @kbobrovs 
I have refactored property saving in sycl-post-link, could you help review to see if the change aligns with your expectation?
Thank you very much.

Reply:
Hi, @kbobrovs 
Do you have any comments about the latest version of property saving function?
Thank you very much.